odo/backends/avro.py

```diff
@@ -0,0 +1,412 @@
+        """
+        Returns an isntance of avro.datafile.DataFileWriter for writing
+        to a new avro file at `uri`.
+
+        Parameters
+        ----------
+
+        uri : str
+            uri of avro existing, non-empty avro file
+
+        sch : avro.schema.Schema object
+
+        Returns
+        -------
+        avro.datafile.DataFileWriter
+        """
+        rec_writer = io.DatumWriter()
+        try: #Python 2.x API
+            df_writer = datafile.DataFileWriter(
+                open(uri, 'wb'),
+                rec_writer,
+                writers_schema = sch,
+                codec=codec
+            )
+        except TypeError: #Python 3.x API
+            df_writer = datafile.DataFileWriter(
+                open(uri, 'wb'),
+                rec_writer,
+                writer_schema = sch,
+                codec=codec
+            )
+
+        return df_writer
+
+    @property
+    def writer(self):
+        if hasattr(self, '_writer'):
+            return self._writer
+        else:
+            if os.path.exists(self.path) and os.path.getsize(self.path) > 0:
+                df_writer = self._get_append_writer(self.path, self.schema, codec=self.codec)
+            else:
+                df_writer = self._get_new_writer(self.path, self.schema, codec=self.codec)
+            self._writer = df_writer
+        return df_writer
+
+    def flush(self):
+        if hasattr(self, '_writer'):
+            self._writer.close()
+            del(self._writer)
+
+@resource.register('.+\.(avro)')
+def resource_avro(uri, schema=None, codec='null', **kwargs):
+    return AVRO(uri, schema=schema, codec=codec, **kwargs)
+
+@dispatch(schema.RecordSchema)
+def discover_schema(sch):
+    return var * Record([(f.name, discover_schema(f.type)) for f in sch.fields])
+
+@dispatch(schema.UnionSchema)
+def discover_schema(sch):
+    try:
+        types = [s.type for s in sch.schemas]
+        assert "null" in types
+        types.remove("null")
+        assert len(types) == 1
+        return Option(AVRO_TYPE_MAP[types[0]])
+    except AssertionError:
+        raise TypeError("odo supports avro UnionSchema only for nullable fields." + \
+                        "Received {0}".format(str([s.type for s in sch.schemas])))
+
+@dispatch(schema.PrimitiveSchema)
+def discover_schema(sch):
+    return AVRO_TYPE_MAP[sch.type]
+
+@dispatch(schema.MapSchema)
+def discover_schema(sch):
+    # Avro map types always have string keys, see https://avro.apache.org/docs/1.7.7/spec.html#Maps
+    return Map(string, discover_schema(sch.values))
+
+@dispatch(schema.ArraySchema)
+def discover_schema(sch):
+    return var * discover_schema(sch.items)
+
+@dispatch(object)
+def discover_schema(sch):
+    raise TypeError('Unable to discover avro type %r' % type(sch).__name__)
+
+@discover.register(AVRO)
+def discover_avro(f, **kwargs):
+    return discover_schema(f.schema)
+
+def make_avsc_object(ds, name="name", namespace="default", depth=0):
```

Comment:
Could you rename this to be more explicit?  Perhaps something like `make_avro_schema()`, or `avro_schema_from_datashape()`...